atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraWriteEteTest.java

```diff
@@ -0,0 +1,123 @@
+
+public class QosCassandraWriteEteTest extends QosCassandraEteTestSetup {
+
+    @Test
+    public void shouldBeAbleToWriteSmallAmountOfBytesIfDoesNotExceedLimit() {
+        writeNTodosOfSize(1, 100);
+    }
+
+    @Test
+    public void shouldBeAbleToWriteSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> writeNTodosOfSize(1, 100));
+    }
+
+    @Test
+    public void shouldBeAbleToWriteLargeAmountsExceedingTheLimitFirstTime() {
+        writeNTodosOfSize(12, 1_000);
+    }
+
+    @Test
+    public void shouldBeAbleToWriteLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        writeNTodosOfSize(1, 20_000);
+        long firstWriteTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        stopwatch = Stopwatch.createStarted();
+        writeNTodosOfSize(200, 1_000);
+        long secondWriteTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        assertThat(secondWriteTime).isGreaterThan(firstWriteTime);
+        assertThat(secondWriteTime - firstWriteTime).isLessThan(MAX_SOFT_LIMITING_SLEEP_MILLIS);
+    }
+
+    @Test
+    public void shouldNotBeAbleToWriteLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        // Have one quick limit-exceeding write, as the rate-limiter
+        // will let anything pass through until the limit is exceeded.
+        writeNTodosOfSize(1, 100_000);
+
+        assertThatThrownBy(() -> writeNTodosOfSize(200, 10_000))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
+
+    @Test
+    public void writeRateLimitShouldBeRespectedByConcurrentWritingThreads() throws InterruptedException {
+        int oneTodoSizeInBytes = 167;
+
+        int numThreads = 5;
+        int numWritesPerThread = 10;
+
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        List<Future> futures = new ArrayList<>(numThreads);
+
+        long start = System.nanoTime();
+        IntStream.range(0, numThreads).forEach(i ->
+                futures.add(executorService.submit(() -> {
+                    IntStream.range(0, numWritesPerThread).forEach(j -> writeNTodosOfSize(1, 100));
+                    return null;
+                })));
+        executorService.shutdown();
+        Preconditions.checkState(executorService.awaitTermination(30L, TimeUnit.SECONDS),
+                "Read tasks did not finish in 30s");
+        long timeTakenToWriteInSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - start);
+
+        assertThatAllWritesWereSuccessful(futures);
+        long actualWriteRate = (numThreads * numWritesPerThread * oneTodoSizeInBytes) / timeTakenToWriteInSeconds;
+        assertThat(actualWriteRate).isLessThan(
+                writeBytesPerSecond + (writeBytesPerSecond / 10 /* to allow burst time */));
```

Comment:
As with the other test, I'm a bit concerned about the possible precision losses with converting the time to a whole number of seconds, and then integer division here

Reply:
DOne.